File: src/archer/math.rs
use anyhow::{Result, ensure, bail};

use super::config::MarketConfig;
use super::types::{MakerLevel, MAX_LEVELS};

pub fn price_to_ticks(price: f64, config: &MarketConfig) -> Result<u64> {
    ensure!(price.is_finite() && price > 0.0, "invalid price: {price}");
    let ticks_f64 = price * config.price_to_ticks_factor();
    ensure!(ticks_f64 >= 0.5, "price {price} below tick resolution");
    ensure!(ticks_f64 <= u64::MAX as f64, "price overflow");
    let ticks = ticks_f64.round() as u64;
    ensure!(ticks > 0, "price {price} rounds to zero ticks");
    Ok(ticks)
}

#[inline]
pub fn base_lots_to_amount(lots: u64, config: &MarketConfig) -> f64 {
    lots as f64 * config.lots_to_base_factor()
}

#[inline]
pub fn quote_lots_to_amount(lots: u64, config: &MarketConfig) -> f64 {
    lots as f64 * config.lots_to_quote_factor()
}

pub fn base_amount_to_lots(amount: f64, config: &MarketConfig) -> Result<u64> {
    ensure!(amount.is_finite() && amount >= 0.0, "invalid base amount: {amount}");
    if amount == 0.0 {
        return Ok(0);
    }
    let lots = amount * config.base_to_lots_factor();
    ensure!(lots >= 1.0, "base amount {amount} below lot resolution");
    Ok(lots.floor() as u64)
}

pub fn quote_amount_to_lots(amount: f64, config: &MarketConfig) -> Result<u64> {
    ensure!(amount.is_finite() && amount >= 0.0, "invalid quote amount: {amount}");
    if amount == 0.0 {
        return Ok(0);
    }
    let lots = amount * config.quote_to_lots_factor();
    ensure!(lots >= 1.0, "quote amount {amount} below lot resolution");
    Ok(lots.floor() as u64)
}

#[derive(Debug, Clone, Copy)]
pub struct Quote {
    pub price: f64,
    pub size: f64,
}

#[derive(Debug, Clone, Default)]
pub struct TwoSidedQuote {
    pub bids: Vec<Quote>,
    pub asks: Vec<Quote>,
}

impl TwoSidedQuote {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn with_bid(mut self, price: f64, size: f64) -> Self {
        self.bids.push(Quote { price, size });
        self
    }

    pub fn with_ask(mut self, price: f64, size: f64) -> Self {
        self.asks.push(Quote { price, size });
        self
    }
}

#[derive(Debug, Clone)]
pub struct BookUpdate {
    pub new_mid_price_ticks: u64,
    pub bid_levels: Vec<MakerLevel>,
    pub ask_levels: Vec<MakerLevel>,
    pub mid_price_changed: bool,
}

pub fn build_book_update(
    quotes: &TwoSidedQuote,
    current_mid_price_ticks: u64,
    config: &MarketConfig,
    is_lo: bool,
) -> Result<BookUpdate> {
    ensure!(quotes.bids.len() <= MAX_LEVELS, "too many bid levels: {}", quotes.bids.len());
    ensure!(quotes.asks.len() <= MAX_LEVELS, "too many ask levels: {}", quotes.asks.len());

    for i in 1..quotes.bids.len() {
        ensure!(
            quotes.bids[i].price < quotes.bids[i - 1].price,
            "bids not strictly descending at index {i}"
        );
    }
    for i in 1..quotes.asks.len() {
        ensure!(
            quotes.asks[i].price > quotes.asks[i - 1].price,
            "asks not strictly ascending at index {i}"
        );
    }
    if let (Some(bb), Some(ba)) = (quotes.bids.first(), quotes.asks.first()) {
        ensure!(bb.price < ba.price, "crossed book: bid {} >= ask {}", bb.price, ba.price);
    }

    // LO books pin `mid_price_ticks` to 0 and each level's `price_offset_ticks`
    // is its absolute price tick. MM books anchor levels to a moving mid and
    // store signed offsets from it.
    let new_mid_price_ticks = if is_lo {
        0
    } else {
        match (quotes.bids.first(), quotes.asks.first()) {
            (Some(b), Some(a)) => price_to_ticks((b.price + a.price) / 2.0, config)?,
            (Some(b), None) => price_to_ticks(b.price, config)?,
            (None, Some(a)) => price_to_ticks(a.price, config)?,
            (None, None) => bail!("empty quote"),
        }
    };

    // For LO the on-chain mid never moves, so there is never a standalone
    // mid-price update to emit.
    let mid_price_changed = !is_lo && new_mid_price_ticks != current_mid_price_ticks;
    let anchor_ticks = new_mid_price_ticks as i64;

    let mut bid_levels: Vec<MakerLevel> = Vec::with_capacity(quotes.bids.len());
    for (i, q) in quotes.bids.iter().enumerate() {
        let price_ticks = price_to_ticks(q.price, config)?;
        let offset = price_ticks as i64 - anchor_ticks;
        let size_lots = base_amount_to_lots(q.size, config)?;

        if i > 0 {
            ensure!(
                offset < bid_levels[i - 1].price_offset_ticks,
                "duplicate bid tick offset at level {i}"
            );
        }

        bid_levels.push(MakerLevel {
            size_in_base_lots: size_lots,
            price_offset_ticks: offset,
        });
    }

    let mut ask_levels: Vec<MakerLevel> = Vec::with_capacity(quotes.asks.len());
    for (i, q) in quotes.asks.iter().enumerate() {
        let price_ticks = price_to_ticks(q.price, config)?;
        let offset = price_ticks as i64 - anchor_ticks;
        let size_lots = base_amount_to_lots(q.size, config)?;

        if i > 0 {
            ensure!(
                offset > ask_levels[i - 1].price_offset_ticks,
                "duplicate ask tick offset at level {i}"
            );
        }

        ask_levels.push(MakerLevel {
            size_in_base_lots: size_lots,
            price_offset_ticks: offset,
        });
    }

    Ok(BookUpdate {
        new_mid_price_ticks,
        bid_levels,
        ask_levels,
        mid_price_changed,
    })
}

File: ts/src/archer/math.ts
// Port of src/archer/math.rs — price/amount conversions and the book-update
// builder. Lots and ticks are bigint (u64/i64 on-chain); prices and sizes are
// number (f64), matching where Rust uses f64.

import type { MakerLevel } from "../generated/types/makerLevel";
import type { MarketConfig } from "./marketConfig";

export const MAX_LEVELS = 16;

const U64_MAX = 18_446_744_073_709_551_615n;

export function priceToTicks(price: number, config: MarketConfig): bigint {
  if (!Number.isFinite(price) || price <= 0) {
    throw new Error(`invalid price: ${price}`);
  }
  const ticksF64 = price * config.priceToTicksFactor;
  if (ticksF64 < 0.5) {
    throw new Error(`price ${price} below tick resolution`);
  }
  if (ticksF64 > Number(U64_MAX)) {
    throw new Error("price overflow");
  }
  const ticks = BigInt(Math.round(ticksF64));
  if (ticks <= 0n) {
    throw new Error(`price ${price} rounds to zero ticks`);
  }
  return ticks;
}

export function baseLotsToAmount(lots: bigint, config: MarketConfig): number {
  return Number(lots) * config.lotsToBaseFactor;
}

export function quoteLotsToAmount(lots: bigint, config: MarketConfig): number {
  return Number(lots) * config.lotsToQuoteFactor;
}

export function baseAmountToLots(amount: number, config: MarketConfig): bigint {
  if (!Number.isFinite(amount) || amount < 0) {
    throw new Error(`invalid base amount: ${amount}`);
  }
  if (amount === 0) {
    return 0n;
  }
  const lots = amount * config.baseToLotsFactor;
  if (lots < 1) {
    throw new Error(`base amount ${amount} below lot resolution`);
  }
  return BigInt(Math.floor(lots));
}

export function quoteAmountToLots(amount: number, config: MarketConfig): bigint {
  if (!Number.isFinite(amount) || amount < 0) {
    throw new Error(`invalid quote amount: ${amount}`);
  }
  if (amount === 0) {
    return 0n;
  }
  const lots = amount * config.quoteToLotsFactor;
  if (lots < 1) {
    throw new Error(`quote amount ${amount} below lot resolution`);
  }
  return BigInt(Math.floor(lots));
}

export interface Quote {
  price: number;
  size: number;
}

export class TwoSidedQuote {
  bids: Quote[] = [];
  asks: Quote[] = [];

  withBid(price: number, size: number): this {
    this.bids.push({ price, size });
    return this;
  }

  withAsk(price: number, size: number): this {
    this.asks.push({ price, size });
    return this;
  }
}

export interface BookUpdate {
  newMidPriceTicks: bigint;
  bidLevels: MakerLevel[];
  askLevels: MakerLevel[];
  midPriceChanged: boolean;
}

export function buildBookUpdate(
  quotes: TwoSidedQuote,
  currentMidPriceTicks: bigint,
  config: MarketConfig,
  isLo: boolean,
): BookUpdate {
  if (quotes.bids.length > MAX_LEVELS) {
    throw new Error(`too many bid levels: ${quotes.bids.length}`);
  }
  if (quotes.asks.length > MAX_LEVELS) {
    throw new Error(`too many ask levels: ${quotes.asks.length}`);
  }

  for (let i = 1; i < quotes.bids.length; i++) {
    if (!(quotes.bids[i]!.price < quotes.bids[i - 1]!.price)) {
      throw new Error(`bids not strictly descending at index ${i}`);
    }
  }
  for (let i = 1; i < quotes.asks.length; i++) {
    if (!(quotes.asks[i]!.price > quotes.asks[i - 1]!.price)) {
      throw new Error(`asks not strictly ascending at index ${i}`);
    }
  }
  const bb = quotes.bids[0];
  const ba = quotes.asks[0];
  if (bb && ba && !(bb.price < ba.price)) {
    throw new Error(`crossed book: bid ${bb.price} >= ask ${ba.price}`);
  }

  // LO books pin mid_price_ticks to 0 and each level's price_offset_ticks is its
  // absolute price tick. MM books anchor levels to a moving mid and store signed
  // offsets from it.
  let newMidPriceTicks: bigint;
  if (isLo) {
    newMidPriceTicks = 0n;
  } else if (bb && ba) {
    newMidPriceTicks = priceToTicks((bb.price + ba.price) / 2, config);
  } else if (bb) {
    newMidPriceTicks = priceToTicks(bb.price, config);
  } else if (ba) {
    newMidPriceTicks = priceToTicks(ba.price, config);
  } else {
    throw new Error("empty quote");
  }

  const midPriceChanged = !isLo && newMidPriceTicks !== currentMidPriceTicks;
  const anchorTicks = newMidPriceTicks;

  const bidLevels: MakerLevel[] = [];
  for (let i = 0; i < quotes.bids.length; i++) {
    const q = quotes.bids[i]!;
    const priceTicks = priceToTicks(q.price, config);
    const offset = priceTicks - anchorTicks;
    const sizeLots = baseAmountToLots(q.size, config);

    if (i > 0 && !(offset < bidLevels[i - 1]!.priceOffsetTicks)) {
      throw new Error(`duplicate bid tick offset at level ${i}`);
    }
    bidLevels.push({ sizeInBaseLots: sizeLots, priceOffsetTicks: offset });
  }

  const askLevels: MakerLevel[] = [];
  for (let i = 0; i < quotes.asks.length; i++) {
    const q = quotes.asks[i]!;
    const priceTicks = priceToTicks(q.price, config);
    const offset = priceTicks - anchorTicks;
    const sizeLots = baseAmountToLots(q.size, config);

    if (i > 0 && !(offset > askLevels[i - 1]!.priceOffsetTicks)) {
      throw new Error(`duplicate ask tick offset at level ${i}`);
    }
    askLevels.push({ sizeInBaseLots: sizeLots, priceOffsetTicks: offset });
  }

  return { newMidPriceTicks, bidLevels, askLevels, midPriceChanged };
}
